File: concepts/03-reliability/python/circuit_breaker.py
"""
Circuit Breaker Pattern Implementation
Prevents cascading failures by stopping requests to failing services
"""
from typing import Callable, Any, Optional
from datetime import datetime, timedelta
from enum import Enum
import asyncio
import random


class CircuitState(Enum):
    CLOSED = "CLOSED"
    OPEN = "OPEN"
    HALF_OPEN = "HALF_OPEN"


class CircuitBreaker:
    """Circuit breaker to prevent cascading failures"""
    
    def __init__(self, failure_threshold: int = 5, reset_timeout: int = 60,
                 monitoring_window: int = 60):
        self.failure_threshold = failure_threshold
        self.reset_timeout = reset_timeout  # seconds
        self.monitoring_window = monitoring_window  # seconds
        
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time: Optional[datetime] = None
        self.next_attempt_time: Optional[datetime] = None
        
        self.failures = []
        self.successes = []
    
    async def execute(self, fn: Callable, *args, **kwargs) -> Any:
        """Execute function with circuit breaker protection"""
        # Check if circuit should transition
        self._check_state()
        
        if self.state == CircuitState.OPEN:
            if self.next_attempt_time and datetime.now() < self.next_attempt_time:
                raise Exception('Circuit breaker is OPEN')
            else:
                # Transition to HALF_OPEN
                self.state = CircuitState.HALF_OPEN
                self.success_count = 0
                print('Circuit breaker: OPEN -> HALF_OPEN')
        
        try:
            result = await fn(*args, **kwargs) if asyncio.iscoroutinefunction(fn) else fn(*args, **kwargs)
            self._on_success()
            return result
        except Exception as error:
            self._on_failure()
            raise error
    
    def _on_success(self):
        """Handle successful request"""
        now = datetime.now()
        self.successes.append(now)
        self._clean_old_records(now)
        
        if self.state == CircuitState.HALF_OPEN:
            self.success_count += 1
            if self.success_count >= 2:
                # Close circuit after 2 successes
                self.state = CircuitState.CLOSED
                self.failure_count = 0
                self.success_count = 0
                print('Circuit breaker: HALF_OPEN -> CLOSED')
        elif self.state == CircuitState.CLOSED:
            # Reset failure count on success
            self.failure_count = 0
    
    def _on_failure(self):
        """Handle failed request"""
        now = datetime.now()
        self.failures.append(now)
        self._clean_old_records(now)
        
        self.failure_count += 1
        self.last_failure_time = now
        
        if self.state == CircuitState.HALF_OPEN:
            # Any failure in HALF_OPEN opens circuit
            self.state = CircuitState.OPEN
            self.next_attempt_time = now + timedelta(seconds=self.reset_timeout)
            print('Circuit breaker: HALF_OPEN -> OPEN')
        elif self.state == CircuitState.CLOSED:
            if self.failure_count >= self.failure_threshold:
                # Open circuit
                self.state = CircuitState.OPEN
                self.next_attempt_time = now + timedelta(seconds=self.reset_timeout)
                print('Circuit breaker: CLOSED -> OPEN')
    
    def _check_state(self):
        """Check and update circuit state"""
        if self.state == CircuitState.OPEN:
            if self.next_attempt_time and datetime.now() >= self.next_attempt_time:
                self.state = CircuitState.HALF_OPEN
                self.success_count = 0
                print('Circuit breaker: OPEN -> HALF_OPEN (timeout expired)')
    
    def _clean_old_records(self, now: datetime):
        """Clean old failure/success records"""
        cutoff = now - timedelta(seconds=self.monitoring_window)
        self.failures = [t for t in self.failures if t > cutoff]
        self.successes = [t for t in self.successes if t > cutoff]
    
    def get_state(self) -> dict:
        """Get current state"""
        return {
            'state': self.state.value,
            'failure_count': self.failure_count,
            'success_count': self.success_count,
            'next_attempt_time': self.next_attempt_time.isoformat() if self.next_attempt_time else None
        }
    
    def get_stats(self) -> dict:
        """Get statistics"""
        now = datetime.now()
        window_start = now - timedelta(seconds=self.monitoring_window)
        
        recent_failures = len([t for t in self.failures if t > window_start])
        recent_successes = len([t for t in self.successes if t > window_start])
        total = recent_failures + recent_successes
        failure_rate = (recent_failures / total * 100) if total > 0 else 0
        
        return {
            'state': self.state.value,
            'failure_count': self.failure_count,
            'success_count': self.success_count,
            'recent_failures': recent_failures,
            'recent_successes': recent_successes,
            'failure_rate': f'{failure_rate:.2f}%',
            'next_attempt_time': self.next_attempt_time.isoformat() if self.next_attempt_time else None
        }
    
    def reset(self):
        """Reset circuit breaker"""
        self.state = CircuitState.CLOSED
        self.failure_count = 0
        self.success_count = 0
        self.last_failure_time = None
        self.next_attempt_time = None
        self.failures = []
        self.successes = []
        print('Circuit breaker: RESET -> CLOSED')


class RetryWithBackoff:
    """Retry with exponential backoff"""
    
    def __init__(self, max_retries: int = 3, base_delay: int = 1000,
                 max_delay: int = 30000, jitter: bool = True):
        self.max_retries = max_retries
        self.base_delay = base_delay  # milliseconds
        self.max_delay = max_delay
        self.jitter = jitter
    
    async def execute(self, fn: Callable, *args, **kwargs) -> Any:
        """Execute function with retry"""
        last_error = None
        
        for attempt in range(self.max_retries + 1):
            try:
                if asyncio.iscoroutinefunction(fn):
                    return await fn(*args, **kwargs)
                else:
                    return fn(*args, **kwargs)
            except Exception as error:
                last_error = error
                
                if attempt < self.max_retries:
                    delay = self._calculate_delay(attempt)
                    print(f"Attempt {attempt + 1} failed, retrying in {delay}ms...")
                    await asyncio.sleep(delay / 1000)
        
        raise last_error
    
    def _calculate_delay(self, attempt: int) -> int:
        """Calculate delay with exponential backoff"""
        # Exponential backoff: baseDelay * 2^attempt
        delay = self.base_delay * (2 ** attempt)
        
        # Cap at maxDelay
        delay = min(delay, self.max_delay)
        
        # Add jitter (random variation)
        if self.jitter:
            jitter_amount = delay * 0.1  # 10% jitter
            delay += random.uniform(-jitter_amount, jitter_amount)
        
        return max(0, int(delay))


class ResilientService:
    """Combined Circuit Breaker + Retry"""
    
    def __init__(self, circuit_breaker_options: dict = None, retry_options: dict = None):
        self.circuit_breaker = CircuitBreaker(**(circuit_breaker_options or {}))
        self.retry = RetryWithBackoff(**(retry_options or {}))
    
    async def call(self, fn: Callable, *args, **kwargs) -> Any:
        """Call function with circuit breaker and retry"""
        return await self.circuit_breaker.execute(
            lambda: self.retry.execute(fn, *args, **kwargs)
        )
    
    def get_stats(self) -> dict:
        """Get statistics"""
        return self.circuit_breaker.get_stats()


async def demonstrate_circuit_breaker():
    """Demonstrate circuit breaker"""
    # Simulate a flaky service
    call_count = 0
    
    async def flaky_service():
        nonlocal call_count
        call_count += 1
        # Fail first 7 calls, then succeed
        if call_count <= 7:
            raise Exception('Service unavailable')
        return {'data': 'Success', 'call_count': call_count}
    
    circuit_breaker = CircuitBreaker(
        failure_threshold=5,
        reset_timeout=5  # 5 seconds
    )
    
    print('=== Circuit Breaker Demonstration ===\n')
    
    # Make requests that will fail
    for i in range(1, 8):
        try:
            await circuit_breaker.execute(flaky_service)
        except Exception as e:
            print(f"Request {i}: {e}")
            print(f"State: {circuit_breaker.get_state()['state']}")
    
    # Circuit should be OPEN now
    print('\nCircuit is OPEN, requests will be rejected immediately')
    try:
        await circuit_breaker.execute(flaky_service)
    except Exception as e:
        print(f"Request rejected: {e}")
    
    # Wait for reset timeout
    print('\nWaiting for reset timeout...')
    await asyncio.sleep(6)
    
    # Circuit should transition to HALF_OPEN
    print('\nCircuit is HALF_OPEN, testing service...')
    try:
        result = await circuit_breaker.execute(flaky_service)
        print(f"Request succeeded: {result}")
    except Exception as e:
        print(f"Request failed: {e}")
    
    # Second success should close circuit
    try:
        result = await circuit_breaker.execute(flaky_service)
        print(f"Request succeeded: {result}")
        print(f"State: {circuit_breaker.get_state()['state']}")
    except Exception as e:
        print(f"Request failed: {e}")
    
    print('\n=== Statistics ===')
    print(circuit_breaker.get_stats())


async def demonstrate_retry():
    """Demonstrate retry with exponential backoff"""
    print('\n=== Retry with Exponential Backoff ===\n')
    
    attempt = 0
    
    async def flaky_service():
        nonlocal attempt
        attempt += 1
        if attempt < 3:
            raise Exception(f'Service error (attempt {attempt})')
        return {'data': 'Success', 'attempt': attempt}
    
    retry = RetryWithBackoff(max_retries=3, base_delay=1000)
    
    try:
        result = await retry.execute(flaky_service)
        print(f'Final result: {result}')
    except Exception as e:
        print(f'All retries exhausted: {e}')


if __name__ == '__main__':
    asyncio.run(demonstrate_circuit_breaker())
    asyncio.run(demonstrate_retry())

File: concepts/03-reliability/node/circuit-breaker.js
/**
 * Circuit Breaker Pattern Implementation
 * Prevents cascading failures by stopping requests to failing services
 */

class CircuitBreaker {
  constructor(options = {}) {
    this.failureThreshold = options.failureThreshold || 5;
    this.resetTimeout = options.resetTimeout || 60000; // 60 seconds
    this.monitoringWindow = options.monitoringWindow || 60000; // 60 seconds
    
    this.state = 'CLOSED'; // CLOSED, OPEN, HALF_OPEN
    this.failureCount = 0;
    this.successCount = 0;
    this.lastFailureTime = null;
    this.nextAttemptTime = null;
    
    this.failures = [];
    this.successes = [];
  }

  /**
   * Execute function with circuit breaker protection
   */
  async execute(fn, ...args) {
    // Check if circuit should transition
    this.checkState();
    
    if (this.state === 'OPEN') {
      if (Date.now() < this.nextAttemptTime) {
        throw new Error('Circuit breaker is OPEN');
      } else {
        // Transition to HALF_OPEN
        this.state = 'HALF_OPEN';
        this.successCount = 0;
        console.log('Circuit breaker: OPEN -> HALF_OPEN');
      }
    }
    
    try {
      const result = await fn(...args);
      this.onSuccess();
      return result;
    } catch (error) {
      this.onFailure();
      throw error;
    }
  }

  /**
   * Handle successful request
   */
  onSuccess() {
    const now = Date.now();
    this.successes.push(now);
    this.cleanOldRecords(now);
    
    if (this.state === 'HALF_OPEN') {
      this.successCount++;
      if (this.successCount >= 2) {
        // Close circuit after 2 successes
        this.state = 'CLOSED';
        this.failureCount = 0;
        this.successCount = 0;
        console.log('Circuit breaker: HALF_OPEN -> CLOSED');
      }
    } else if (this.state === 'CLOSED') {
      // Reset failure count on success
      this.failureCount = 0;
    }
  }

  /**
   * Handle failed request
   */
  onFailure() {
    const now = Date.now();
    this.failures.push(now);
    this.cleanOldRecords(now);
    
    this.failureCount++;
    this.lastFailureTime = now;
    
    if (this.state === 'HALF_OPEN') {
      // Any failure in HALF_OPEN opens circuit
      this.state = 'OPEN';
      this.nextAttemptTime = now + this.resetTimeout;
      console.log('Circuit breaker: HALF_OPEN -> OPEN');
    } else if (this.state === 'CLOSED') {
      if (this.failureCount >= this.failureThreshold) {
        // Open circuit
        this.state = 'OPEN';
        this.nextAttemptTime = now + this.resetTimeout;
        console.log('Circuit breaker: CLOSED -> OPEN');
      }
    }
  }

  /**
   * Check and update circuit state
   */
  checkState() {
    if (this.state === 'OPEN' && Date.now() >= this.nextAttemptTime) {
      this.state = 'HALF_OPEN';
      this.successCount = 0;
      console.log('Circuit breaker: OPEN -> HALF_OPEN (timeout expired)');
    }
  }

  /**
   * Clean old failure/success records
   */
  cleanOldRecords(now) {
    const cutoff = now - this.monitoringWindow;
    this.failures = this.failures.filter(time => time > cutoff);
    this.successes = this.successes.filter(time => time > cutoff);
  }

  /**
   * Get current state
   */
  getState() {
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      nextAttemptTime: this.nextAttemptTime
    };
  }

  /**
   * Get statistics
   */
  getStats() {
    const now = Date.now();
    const windowStart = now - this.monitoringWindow;
    
    const recentFailures = this.failures.filter(time => time > windowStart).length;
    const recentSuccesses = this.successes.filter(time => time > windowStart).length;
    const total = recentFailures + recentSuccesses;
    const failureRate = total > 0 ? (recentFailures / total) * 100 : 0;
    
    return {
      state: this.state,
      failureCount: this.failureCount,
      successCount: this.successCount,
      recentFailures,
      recentSuccesses,
      failureRate: failureRate.toFixed(2) + '%',
      nextAttemptTime: this.nextAttemptTime ? new Date(this.nextAttemptTime).toISOString() : null
    };
  }

  /**
   * Reset circuit breaker
   */
  reset() {
    this.state = 'CLOSED';
    this.failureCount = 0;
    this.successCount = 0;
    this.lastFailureTime = null;
    this.nextAttemptTime = null;
    this.failures = [];
    this.successes = [];
    console.log('Circuit breaker: RESET -> CLOSED');
  }
}

/**
 * Retry with exponential backoff
 */
class RetryWithBackoff {
  constructor(options = {}) {
    this.maxRetries = options.maxRetries || 3;
    this.baseDelay = options.baseDelay || 1000; // 1 second
    this.maxDelay = options.maxDelay || 30000; // 30 seconds
    this.jitter = options.jitter !== false; // Enable jitter by default
  }

  async execute(fn, ...args) {
    let lastError;
    
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      try {
        return await fn(...args);
      } catch (error) {
        lastError = error;
        
        if (attempt < this.maxRetries) {
          const delay = this.calculateDelay(attempt);
          console.log(`Attempt ${attempt + 1} failed, retrying in ${delay}ms...`);
          await this.sleep(delay);
        }
      }
    }
    
    throw lastError;
  }

  calculateDelay(attempt) {
    // Exponential backoff: baseDelay * 2^attempt
    let delay = this.baseDelay * Math.pow(2, attempt);
    
    // Cap at maxDelay
    delay = Math.min(delay, this.maxDelay);
    
    // Add jitter (random variation)
    if (this.jitter) {
      const jitterAmount = delay * 0.1; // 10% jitter
      delay += (Math.random() * 2 - 1) * jitterAmount;
    }
    
    return Math.max(0, Math.floor(delay));
  }

  sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
  }
}

/**
 * Combined Circuit Breaker + Retry
 */
class ResilientService {
  constructor(options = {}) {
    this.circuitBreaker = new CircuitBreaker(options.circuitBreaker);
    this.retry = new RetryWithBackoff(options.retry);
  }

  async call(fn, ...args) {
    return this.circuitBreaker.execute(
      () => this.retry.execute(fn, ...args),
      ...args
    );
  }

  getStats() {
    return this.circuitBreaker.getStats();
  }
}

// Example usage
async function demonstrateCircuitBreaker() {
  // Simulate a flaky service
  let callCount = 0;
  const flakyService = async () => {
    callCount++;
    // Fail first 7 calls, then succeed
    if (callCount <= 7) {
      throw new Error('Service unavailable');
    }
    return { data: 'Success', callCount };
  };

  const circuitBreaker = new CircuitBreaker({
    failureThreshold: 5,
    resetTimeout: 5000 // 5 seconds
  });

  console.log('=== Circuit Breaker Demonstration ===\n');

  // Make requests that will fail
  for (let i = 1; i <= 7; i++) {
    try {
      await circuitBreaker.execute(flakyService);
    } catch (error) {
      console.log(`Request ${i}: ${error.message}`);
      console.log(`State: ${circuitBreaker.getState().state}`);
    }
  }

  // Circuit should be OPEN now
  console.log('\nCircuit is OPEN, requests will be rejected immediately');
  try {
    await circuitBreaker.execute(flakyService);
  } catch (error) {
    console.log(`Request rejected: ${error.message}`);
  }

  // Wait for reset timeout
  console.log('\nWaiting for reset timeout...');
  await new Promise(resolve => setTimeout(resolve, 6000));

  // Circuit should transition to HALF_OPEN
  console.log('\nCircuit is HALF_OPEN, testing service...');
  try {
    const result = await circuitBreaker.execute(flakyService);
    console.log(`Request succeeded:`, result);
  } catch (error) {
    console.log(`Request failed: ${error.message}`);
  }

  // Second success should close circuit
  try {
    const result = await circuitBreaker.execute(flakyService);
    console.log(`Request succeeded:`, result);
    console.log(`State: ${circuitBreaker.getState().state}`);
  } catch (error) {
    console.log(`Request failed: ${error.message}`);
  }

  console.log('\n=== Statistics ===');
  console.log(circuitBreaker.getStats());
}

// Demonstrate retry with exponential backoff
async function demonstrateRetry() {
  console.log('\n=== Retry with Exponential Backoff ===\n');

  let attempt = 0;
  const flakyService = async () => {
    attempt++;
    if (attempt < 3) {
      throw new Error(`Service error (attempt ${attempt})`);
    }
    return { data: 'Success', attempt };
  };

  const retry = new RetryWithBackoff({
    maxRetries: 3,
    baseDelay: 1000
  });

  try {
    const result = await retry.execute(flakyService);
    console.log('Final result:', result);
  } catch (error) {
    console.log('All retries exhausted:', error.message);
  }
}

if (require.main === module) {
  demonstrateCircuitBreaker()
    .then(() => demonstrateRetry())
    .catch(console.error);
}

module.exports = {
  CircuitBreaker,
  RetryWithBackoff,
  ResilientService
};

